json-schema/json-schema/src/main/scala/endpoints4s/Validated.scala

```diff
@@ -110,4 +112,60 @@ object Validated {
       case Failure(error) => Invalid(error.getMessage)
     }
 
+  /** Turns an `IterableOnce[Validated[A]]`into a `Validated[IterableOnce[A]]`.
+    * If all values in the sequence are `Valid`, the result will be `Valid`
+    * If there is at least one `Invalid`, the result will be an `Invalid` containing the error messages of all `Invalid`s
+    * If the `Seq` is empty, will return a `Valid[Seq.empty]`.
+    *
+    * @see [[scala.concurrent.Future.sequence]]
+    *
+    * @param in   the `IterableOnce` of Validateds which will be sequenced
+    * @tparam A   the type of the value inside the Validateds
+    * @tparam CC  the type of the `IterableOnce` of Validateds
+    * @tparam To  the type of the resulting collection
+    * @return     the `Validated` of the resulting collection
+    */
+  def sequence[A, CC[+X] <: IterableOnce[X], To](
+      in: CC[Validated[A]]
+  )(implicit bf: BuildFrom[CC[Validated[A]], A, To]): Validated[To] = {
+    val builder = bf.newBuilder(in)
+
+    in.iterator
+      .foldLeft[Validated[Unit]](Valid(())) { case (folded, next) =>
+        folded.zip(next).map { n => builder += n; () }
+      }
+      .map(_ => builder.result())
+
+  }
+
+  /** Transforms a `IterableOnce[A]` into a `Validated[IterableOnce[B]]`
+    *  using the provided function `A => Validated[B]`.
+    *  This is useful for validating all entries in an IterableOnce:
+    *
+    *  {{{
+    *    val myValidatedList = Validated.traverse(myList)(x => myValidateFunc(x))
+    *  }}}
+    *
+    * @see [[scala.concurrent.Future.traverse]]
+    *
+    * @tparam A        the type of the value inside the collection
+    * @tparam B        the type of the value of the returned Validateds
+    * @tparam M        the type of the collection of Validateds
+    * @param in        the collection to be mapped over with the provided function to produce a collection of Validateds that is then sequenced into a Validated collection
+    * @param fn        the function to be mapped over the collection to produce a collection of Validateds
+    * @return          the Validated of the collection of results
+    */
+  def traverse[A, B, M[X] <: IterableOnce[X]](
```

Comment:
I think this one too should return a `Validated[To]` so that we can return a `Map`, for instance:

~~~ scala
val validatedMaps: Validated[Map[Int, Int]] =
  Validated.traverse(List(1, 2, 3))(x => Valid(x -> x * x))
~~~

Reply:
I just tried, but this doesn't actually work. It returns `Validated[List[(Int, Int)]]`.

I'm not entirely sure, but I think this might be a result of using the Builder, the `zip` only exists to handle the invalid cases, and the map only does the side-effect of adding a value to the mutable builder. That's why I explicitly have the map call return Unit, to clarify that this is what happens. I could've built a Validated.foreach() and use that for the same result.

In my first attempt, the map returned the builder instance itself, but that doesn't change the fact that we depend on a side effect here which means the Tupler can't do anything useful.

Reply:
Ah, OK. Indeed, for a `Map` to be returned we need something more sophisticated.

Hmm, but I believe there is a good reason for having this `To` parameter in `Future.sequence`… I don’t know what is it so maybe it’s simpler to use `CC[_]` everywhere (and to get rid of `To` in `sequence`)?

Reply:
If you mean I should do this:

```
  def sequence[A, CC[_] <: IterableOnce[_]](
      in: CC[Validated[A]]
  )(implicit bf: BuildFrom[CC[Validated[A]], A, CC[A]]): Validated[CC[A]] = {
    val builder = bf.newBuilder(in)

    in.iterator
      .foldLeft[Validated[Unit]](Valid(())) { case (folded, next) =>
        folded.zip(next).map { n => builder += n; () }
      }
      .map(_ => builder.result())
  }
```

This gives the compiler errors:

```
[error] /home/gerret/IdeaProjects/endpoints4s/json-schema/json-schema/src/main/scala/endpoints4s/Validated.scala:133:8: value iterator is not a member of type parameter CC[endpoints4s.Validated[A]]
[error]     in.iterator
[error]        ^
[error] /home/gerret/IdeaProjects/endpoints4s/json-schema/json-schema/src/main/scala/endpoints4s/Validated.scala:135:16: value zip is not a member of Any
[error]         folded.zip(next).map { n => builder += n; () }
[error]                ^
```

Reply:
The error message is confusing but I think if we do:

~~~ diff
- def sequence[A, CC[_] <: IterableOnce[_]]
+ def sequence[A, CC[X] <: IterableOnce[X]]
~~~

That should solve the issue (at least I just tried with Scala 3, I hope this is okay for the other versions of the compiler)

Reply:
Alright, in my latest commit it's now consistent between both functions.